### ISSUE TITLE
.NET 9 - OpenAPI Not Generating Output

### ISSUE THREAD

**Issue by sbwalker** [dotnet/aspnetcore#58805]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

I am trying to integrate OpenAPI into an existing Blazor project which was recently upgraded to .NET 9.

I added the following project reference to the Server project:

```
    <PackageReference Include="Microsoft.AspNetCore.OpenApi" Version="9.0.0-rc.2.24474.3" />
```

I modified the Startup.cs (yes this project has never migrated to the "new" Program.cs approach - but that should not affect the result):

```
        public void ConfigureServices(IServiceCollection services)
        {
            ...
            services.AddMvc();
            services.AddRazorPages();
            services.AddOpenApi();
            ...
        }
```

```
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILogger<Startup> logger)
        {
            ...
            app.UseEndpoints(endpoints =>
            {
                endpoints.MapOpenApi();
            });

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
                endpoints.MapRazorPages();
            });
            ...
        }
```

Note that this application is using traditional Controllers and Razor Pages - not minimal APIs.

When I run the application and browse to /openapi/v1.json I get a blank screen with no output:

![Image](https://github.com/user-attachments/assets/a21471cf-adc4-45d3-9d9f-f693b7db84b7)

The Visual Studio console contains the following information:

```
Microsoft.AspNetCore.Routing.EndpointMiddleware: Information: Executing endpoint 'HTTP: GET /openapi/{documentName}.json'
Microsoft.AspNetCore.Routing.EndpointMiddleware: Information: Executed endpoint 'HTTP: GET /openapi/{documentName}.json'
Exception thrown: 'System.InvalidOperationException' in System.Private.CoreLib.dll
Exception thrown: 'System.InvalidOperationException' in System.Private.CoreLib.dll
Exception thrown: 'System.InvalidOperationException' in System.Private.CoreLib.dll
Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request finished HTTP/1.1 GET http://localhost:44357/openapi/v1.json - 200 - - 981.2280ms
```

### Expected Behavior

Browsing to /openapi/v1.json should display the Open API document 

### Exceptions (if any)

Exception thrown: 'System.InvalidOperationException' in System.Private.CoreLib.dll

### .NET Version

 9.0.0-rc.2.24473.5

### Anything else?

Note that this application already had Swagger integrated previously:

```
            services.AddSwaggerGen(options =>
            {
                options.CustomSchemaIds(type => type.ToString()); // Handle SchemaId already used for different type
            });
            services.TryAddSwagger(_useSwagger);

...

            if (_useSwagger)
            {
                app.UseSwagger();
                app.UseSwaggerUI(c => { c.SwaggerEndpoint("/swagger/" + Constants.Version + "/swagger.json", Constants.PackageId + " " + Constants.Version); });
            }
```

and the Swagger UI still works fine:

![Image](https://github.com/user-attachments/assets/20bfaadb-48d1-4b32-ba39-f914561d0201)

**Comment by sbwalker** [comment]
the actual exception is:

```
Unsupported parameter source: ModelBinding
Microsoft.AspNetCore.OpenApi
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.<GetParametersAsync>d__28.MoveNext()
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.<GetOperationAsync>d__21.MoveNext()
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.<GetOperationsAsync>d__20.MoveNext()
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.<GetOpenApiPathsAsync>d__19.MoveNext()
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.<GetOpenApiDocumentAsync>d__12.MoveNext()
   at Microsoft.AspNetCore.Builder.OpenApiEndpointRouteBuilderExtensions.<>c__DisplayClass0_0.<<MapOpenApi>b__0>d.MoveNext()
   at Microsoft.AspNetCore.Http.Generated.<GeneratedRouteBuilderExtensions_g>F56B68D2B55B5B7B373BA2E4796D897848BC0F04A969B1AF6260183E8B9E0BAF2__GeneratedRouteBuilderExtensionsCore.<>c__DisplayClass2_0.<<MapGet0>g__RequestHandler|5>d.MoveNext()
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<<Invoke>g__AwaitRequestTask|7_0>d.MoveNext()
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.<Invoke>d__5.MoveNext()
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.<Invoke>d__6.MoveNext()
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.<Invoke>d__11.MoveNext()
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.<Invoke>d__6.MoveNext()
   at Oqtane.Infrastructure.JwtMiddleware.<Invoke>d__2.MoveNext() 
   in C:\Source\Projects\oqtane.framework\Oqtane.Server\Infrastructure\Middleware\JwtMiddleware.cs:line 88
```

Line 88 in Jwtmiddleware is simply:

```
if (_next != null) await _next(context);
```

**Comment by danroth27** [comment]
@captainsafia @mikekistler

**Comment by captainsafia** [comment]
@sbwalker Are there any controller actions in your application that take advantage of custom model binding from the route and/or query parameter as opposed to the body?

**Comment by sbwalker** [comment]
Yes there are controller methods that rely on query parameters ie.

```
        // GET: api/<controller>?siteid=x
        [HttpGet]
        public IEnumerable<Page> Get(string siteid)
```

**Comment by captainsafia** [comment]
@sbwalker Those should be totally fine since you're just using the `string` type which uses the standard binding rules.

Is there a chance that you're able to provide a repro for this? My hunch is that you're seeing the same issue as https://github.com/dotnet/aspnetcore/issues/59013.

**Comment by trevonmckay** [comment]
I too am not getting any output, just ERR_EMPTY_RESPONSE. I don't see any exceptions in the output log though.

```csharp
services
    .AddControllers(options =>
    {
        options.Conventions.Add(new ContentNegotiationActionModelConvention());
        options.InputFormatters.Insert(0, JsonPatchExtensions.GetJsonPatchInputFormatter());
        options.ReturnHttpNotAcceptable = false;
    })
    .AddJsonOptions(options =>
    {
        options.JsonSerializerOptions.Converters.Add(new TypeDiscriminatingConverter());
        options.JsonSerializerOptions.Converters.Add(new JsonStringEnumConverter(JsonNamingPolicy.CamelCase));
        options.JsonSerializerOptions.Converters.Add(new NonPublicConstructorJsonConverterFactory());
        options.JsonSerializerOptions.Converters.Add(new JTokenConverterFactory());
        options.JsonSerializerOptions.Converters.Add(new JsonPatchDocumentConverterFactory());
        options.JsonSerializerOptions.DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull;
        options.JsonSerializerOptions.ReferenceHandler = ReferenceHandler.IgnoreCycles;
        options.JsonSerializerOptions.PropertyNamingPolicy = JsonNamingPolicy.CamelCase;

        options.JsonSerializerOptions.MakeReadOnly();
    });

services.AddOpenApi();
```

```csharp
app.UseEndpoints(endpoints =>
{
    endpoints.MapOpenApi("/openapi/api.json");

    endpoints.MapControllers();

    endpoints.MapHealthChecks(
        "/healthz/live",
        new Microsoft.AspNetCore.Diagnostics.HealthChecks.HealthCheckOptions
        {
            Predicate = (_) => false,
        })
    .ShortCircuit();

    endpoints.MapHealthChecks(
        "/healthz/ready",
        new Microsoft.AspNetCore.Diagnostics.HealthChecks.HealthCheckOptions
        {
            Predicate = healthCheck => healthCheck.Tags.Contains("ready"),
        },
        TimeSpan.FromSeconds(30));

    endpoints.MapHealthChecks(
        "/health",
        new Microsoft.AspNetCore.Diagnostics.HealthChecks.HealthCheckOptions
        {
            ResponseWriter = UIResponseWriter.WriteHealthCheckUIResponse,
        });
});
```

![Image](https://github.com/user-attachments/assets/443cae12-f74c-48ca-bbb6-ece47157abcc)

**Comment by trevonmckay** [comment]
After disabling Just My Code I am now seeing the exception 

> System.InvalidOperationException: 'Cannot resolve scoped service 'Microsoft.Extensions.Options.IOptionsSnapshot`1[Microsoft.AspNetCore.OpenApi.OpenApiOptions]' from root provider.'

**Comment by trevonmckay** [comment]
I add this just to test and I still receive the "Cannot resolve scoped service 'Microsoft.Extensions.Options.IOptionsSnapshot`1[Microsoft.AspNetCore.OpenApi.OpenApiOptions]" error

```csharp
services.AddOpenApi();

services.AddScoped<IOptionsSnapshot<OpenApiOptions>>(sp =>
{
    OpenApiOptions options = new();
    return new StaticOptions<OpenApiOptions>(options);
});
```

**Comment by BoerG-RDW** [comment]
I am facing the same issue as @trevonmckay . I eventually found out that when I set the ASPNETCORE_ENVIRONMENT to `Development`, you get the exception. If you set it to `Production`, everything runs fine.

**Comment by mikekistler** [comment]
I tried recreating this problem and was unsuccessful. What I did:

- dotnet new blazorserver
- dotnet add package Microsoft.AspNetCore.OpenApi
- Added the two lines in Program.cs to add OpenApi services and map the openapi endpoint.

Running this project produces a (rather empty) OpenAPI document at /openapi/v1.json.

Is there something I can do in this project to recreate the problem or do I need to take a completely different path?

**Comment by danroth27** [comment]
@trevonmckay @BoerG-RDW Can you provide a simplified GitHub repo that reproduces the problem you're seeing?

**Comment by BoerG-RDW** [comment]
really strange. I'm still not sure what is causing it, but I threw out everything that I didn't need, made a copy of only the source code + project/solution file. The copy worked fine. Finally had a look at the hidden directories and when I threw out the .vs directory, it started working suddenly. It's mostly binary stuff inside that directory, so I can't really see what the actual problem is, but throwing out that directory fixed the issue for me. Visual Studio will recreate it automatically, so I don't think you'll loose anything if you throw it out.

**Comment by trevonmckay** [comment]
I've managed to get this working now with two changes

1.
```csharp
public class StaticOptions<TOptions> : IOptionsSnapshot<TOptions>
    where TOptions : class
{
    private readonly TOptions _options;

    public StaticOptions(TOptions options)
    {
        _options = options;
    }

    public TOptions Value
    {
        get
        {
            return _options;
        }
    }

    public TOptions Get(string name)
    {
        return _options;
    }
}
```

Manually adding the IOptionsSnapshot<OpenApiOptions> as a singleton
```csharp
services.AddSingleton<IOptionsSnapshot<OpenApiOptions>>(sp =>
{
    OpenApiOptions options = new();
    return new StaticOptions<OpenApiOptions>(options);
});

services.AddOpenApi();
```
which resolved the "Cannot resolve scoped service 'Microsoft.Extensions.Options.IOptionsSnapshot`1[Microsoft.AspNetCore.OpenApi.OpenApiOptions]" error. 

That revealed another error, the JSON MaxDepth was exceeded during the document generation. Increasing the max depth to 256 seemed to do the trick for that.

```csharp
services.Configure<JsonOptions>(jsonOptions =>
{
    jsonOptions.SerializerOptions.DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull;
    jsonOptions.SerializerOptions.MaxDepth = 256;
    jsonOptions.SerializerOptions.PropertyNamingPolicy = JsonNamingPolicy.CamelCase;
    jsonOptions.SerializerOptions.ReferenceHandler = ReferenceHandler.IgnoreCycles;
});
```

**Comment by captainsafia** [comment]
> [@danroth27](https://github.com/danroth27) I have provided some repro steps:
> 
> 1. Clone https://github.com/oqtane/oqtane.framework (open source project)
> 2. Add the following project reference to Oqtane.Server\Oqtane.Server.csproj
> 
> ```
>     <PackageReference Include="Microsoft.AspNetCore.OpenApi" Version="9.0.0" />
> ```
> 
>  3. Modify Oqtane.Server\Startup.cs to add OpenApi:
> 
> ```
>         public void ConfigureServices(IServiceCollection services)
>         {
>             ...
>             services.AddRazorPages();
>             services.AddOpenApi();
>             ...
>         }
> ```
> 
>  ```
>         public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILogger<Startup> logger)
>         {
>             ...
>             app.UseEndpoints(endpoints =>
>             {
>                 endpoints.MapOpenApi();
>             });
> 
>             app.UseEndpoints(endpoints =>
>             {
>                 endpoints.MapControllers();
>                 endpoints.MapRazorPages();
>             });
>             ...
>         }
> ```
> 
>  4. Rebuild Solution / Run
> 
> Choose SQLite as the Database option Configure Administrator Account (Username: host / Password: P@ssw0rd / Email: [host@host.com](mailto:host@host.com) / Default Site Template) Install
> 
> 5. Browse to http://localhost:44357/openapi/v1.json
> 
> You will see a blank screen with no output:
> 
> ![Image](https://github.com/user-attachments/assets/a21471cf-adc4-45d3-9d9f-f693b7db84b7)

I was able to repro with these steps. It looks like what you're seeing is indeed the same issue as https://github.com/dotnet/aspnetcore/issues/59013 which was resolved in https://github.com/dotnet/aspnetcore/pull/59035. I can open a backport for this issue to .NET 9.

@BoerG-RDW @trevonmckay The issue you are seeing are distinct from the Oqtane issue that is reported here. I'd recommend filing another issue for that.

In the meantime, I'll close this as a dupe of https://github.com/dotnet/aspnetcore/issues/59013 and track a backport for that.

**Comment by wolfgang-hartl** [comment]
Hello everyone,

> TLDR: I'm facing a Problem when separating endpoints outside of Program.cs where they don't get included in the generated openapi.json

I’m not sure where this should be posted or where it belongs. I’ll write it here for now, but please let me know if I should post it somewhere else.

I’m also having issues with OpenAPI generation. I’m using a minimal API and just experimenting with the .NET framework. I’ve followed the approach described in the documentation [here](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/route-handlers?view=aspnetcore-9.0#endpoint-defined-outside-of-programcs) and moved my route handlers into separate files.

```
using APIRoutes;

var builder = WebApplication.CreateBuilder(args);
builder.Services.AddOpenApi();

var app = builder.Build();
app.MapGet("/", () => "Hello World!");
app.MapGet("/hello/{id}", (int id) => "Hello World!");

TodoEndpoints.Map(app);

if (app.Environment.IsDevelopment())
{
    app.MapOpenApi();
}



app.Run();
```
However, OpenAPI does not recognize these routes and only generates the JSON file with the routes registered in program.cs. Could this be related to the bug mentioned here?

That's how the generated JSON File looks like: 

```
{
  "openapi": "3.0.1",
  "info": {
    "title": "api-dotnet | v1",
    "version": "1.0.0"
  },
  "servers": [
    {
      "url": "http://localhost:5190"
    }
  ],
  "paths": {
    "/": {
      "get": {
        "tags": [
          "api-dotnet"
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "text/plain": {
                "schema": {
                  "type": "string"
                }
              }
            }
          }
        }
      }
    },
    "/hello/{id}": {
      "get": {
        "tags": [
          "api-dotnet"
        ],
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "integer",
              "format": "int32"
            }
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "text/plain": {
                "schema": {
                  "type": "string"
                }
              }
            }
          }
        }
      }
    }
  },
  "components": { },
  "tags": [
    {
      "name": "api-dotnet"
    }
  ]
}
´´´

**Comment by captainsafia** [comment]
@wolfgang-hartl Your best bet is to file an new issue with a pointer to your sample code (https://github.com/dotnet/aspnetcore/issues/new). In particular, the implementation for `TodoEndpoints.Map`.

**Comment by wolfgang-hartl** [comment]
@captainsafia that was fast! :-) Thanks for your response - i will do that. The Implementation is just a copy/paste from the docs mentioned, but i will write that all together and create a new issue 👍

**Comment by raphaelschmitz00** [comment]
As someone who faced the same symptom while trying to switch from SwashBuckle:
I had `[ApiExplorerSettings]` attributes on all my controllers, and if they have that, they don't show up either.

**Comment by sbwalker** [comment]
@raphaelschmitz00 have you tried this on .NET 10 preview releases? Has it been fixed?

**Comment by fededim** [comment]
**I managed to get it working in NET9 using a hybrid Microsoft OpenAPI and SwaggerUI**

- **Add JWT in DI**

```csharp
            builder.Services.AddAuthentication(opt =>
            {
                opt.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
                opt.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
                opt.DefaultScheme = JwtBearerDefaults.AuthenticationScheme;
            })
            .AddJwtBearer(options =>
            {
                options.TokenValidationParameters = new TokenValidationParameters
                {
                    ValidateIssuer = true,
                    ValidateAudience = true,
                    ValidateLifetime = true,
                    ValidateIssuerSigningKey = true,
                    ... whatever
                };
            });

```

- **Add Microsoft OpenAPI in DI**
_The part with comment MAGIC CODE needs to be added make a fully functional Swagger UI_

```csharp
            builder.Services.AddOpenApi(options =>
            {
                options.AddDocumentTransformer((document, context, cancellationToken) =>
                {
                    document.Info = new OpenApiInfo
                    {
                        Title = $"AIAffiliateMarketingAPI ({builder.Environment.EnvironmentName})",
                        Version = "v1",
                        Description = "An AI affiliate marketing api associated to AIAffiliateMarketing app"                        
                    };

                    var securitySchemes = new Dictionary<string, OpenApiSecurityScheme>
                    {
                        [JwtBearerDefaults.AuthenticationScheme] = new OpenApiSecurityScheme()
                        {
                            Type = SecuritySchemeType.Http,
                            BearerFormat = "Json Web Token",
                            In = ParameterLocation.Header,
                            Description = "Please insert JWT token into field",
                            Name = "Authorization",
                            Scheme = JwtBearerDefaults.AuthenticationScheme,
                            Reference = new OpenApiReference { Type = ReferenceType.SecurityScheme, Id = JwtBearerDefaults.AuthenticationScheme },
                        }
                    };

                    document.Components ??= new OpenApiComponents();
                    document.Components.SecuritySchemes = securitySchemes;

                    // "MAGIC CODE" taken from https://learn.microsoft.com/en-us/aspnet/core/fundamentals/openapi/customize-openapi?view=aspnetcore-9.0#use-document-transformers
                    // to include JWT requirement inside OpenAPI json
                    // without this Swagger UI does not send authorization header and does not work correctly
                    foreach (var operation in document.Paths.Values.SelectMany(path => path.Operations))
                    {
                        operation.Value.Security.Add(new OpenApiSecurityRequirement
                        {
                            
                            [new OpenApiSecurityScheme { Reference = new OpenApiReference { Id = "Bearer", Type = ReferenceType.SecurityScheme } }] = Array.Empty<string>()
                        });
                    }

                    return Task.CompletedTask;
                });
            });
```


- **Add OpenAPI and Swagger UI in Application**

```csharp
            var app = builder.Build();

            if (app.Environment.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
                app.MapOpenApi();
                app.UseSwaggerUI(c =>
                {
                    c.RoutePrefix = "openapi";
                    c.SwaggerEndpoint("v1.json", $"AIAffiliateMarketingAPI v1");
                });
            }

```

<img width="3839" height="1072" alt="Image" src="https://github.com/user-attachments/assets/fb5fccc0-ca05-403c-9112-c686c66e92db" />
